### DISCUSSION TITLE
Modifying the SID error running salt in manager

### DISCUSSION THREAD

**Discussion by jlaboy71** [Security-Onion-Solutions/securityonion#2629] [General] [answered]
Hi all,

Following the instructions from the [documentation](https://docs.securityonion.net/en/2.3/managing-alerts.html) 

After modifying the file: /opt/so/saltstack/local/pillar/minions/securityonion_standalone.sls

and adding the following disabled SIDs under the idstools section:

dstools:
  config:
    ruleset: 'ETOPEN'
    oinkcode: ''
    urls:
  sids:
    enabled:
    disabled:
        - 2016149 
        - 2016150
    modify:

I am told to run this command to apply:  salt $SENSORNAME_$ROLE state.apply idstools

When I do this, I get this:

[ERROR   ] 
Salt request timed out. The master is not responding. You may need to run your command with `--async` in order to bypass the congested event bus. With `--async`, the CLI tool will print the job id (jid) and exit immediately without listening for responses. You can then use `salt-run jobs.lookup_jid` to look up the results of the job in the job cache later.

When I try to use the --async this is what I get: 

[netadmin@securityonion minions]$ **salt * test.ping**
[ERROR   ] 
Salt request timed out. The master is not responding. You may need to run your command with `--async` in order to bypass the congested event bus. With `--async`, the CLI tool will print the job id (jid) and exit immediately without listening for responses. You can then use `salt-run jobs.lookup_jid` to look up the results of the job in the job cache later.

**[netadmin@securityonion minions]$ salt * test.ping --async**
[ERROR   ] An un-handled exception was caught by salt's global exception handler:
AuthenticationError: Authentication error occurred.
Traceback (most recent call last):
  File "/usr/bin/salt", line 11, in <module>
    load_entry_point('salt==3002.2', 'console_scripts', 'salt')()
  File "/usr/lib/python3.6/site-packages/salt/scripts.py", line 539, in salt_main
    client.run()
  File "/usr/lib/python3.6/site-packages/salt/cli/salt.py", line 151, in run
    jid = self.local_client.cmd_async(**kwargs)
  File "/usr/lib/python3.6/site-packages/salt/client/__init__.py", line 447, in cmd_async
    tgt, fun, arg, tgt_type, ret, jid=jid, kwarg=kwarg, listen=False, **kwargs
  File "/usr/lib/python3.6/site-packages/salt/client/__init__.py", line 348, in run_job
    **kwargs
  File "/usr/lib/python3.6/site-packages/salt/client/__init__.py", line 1873, in pub
    raise AuthenticationError(err_msg)
salt.exceptions.AuthenticationError: Authentication error occurred.
Traceback (most recent call last):
  File "/usr/bin/salt", line 11, in <module>
    load_entry_point('salt==3002.2', 'console_scripts', 'salt')()
  File "/usr/lib/python3.6/site-packages/salt/scripts.py", line 539, in salt_main
    client.run()
  File "/usr/lib/python3.6/site-packages/salt/cli/salt.py", line 151, in run
    jid = self.local_client.cmd_async(**kwargs)
  File "/usr/lib/python3.6/site-packages/salt/client/__init__.py", line 447, in cmd_async
    tgt, fun, arg, tgt_type, ret, jid=jid, kwarg=kwarg, listen=False, **kwargs
  File "/usr/lib/python3.6/site-packages/salt/client/__init__.py", line 348, in run_job
    **kwargs
  File "/usr/lib/python3.6/site-packages/salt/client/__init__.py", line 1873, in pub
    raise AuthenticationError(err_msg)
salt.exceptions.AuthenticationError: Authentication error occurred.
[netadmin@securityonion minions]$ 

Can I get some advice?

Thanks,

Jose

**Comment by m0duspwnens** [comment]
@jlaboy71 
 Can you check the status of the salt-master service?
 `sudo systemctl status salt-master`

 If it isn't running, start it with: `sudo systemctl start salt-master`

 If it is running, restart it with: `sudo systemctl restart salt-master`

 Check the status again:
 `sudo systemctl status salt-master`

 Rerun: `salt $SENSORNAME_$ROLE state.apply idstools`
 It looks like in your case $SENSORNAME_$ROLE would be securityonion_standalone based of the sls file name about that was modified.

Thanks,
Josh

**Reply by jlaboy71** [reply]
Josh,

Thank you for responding and providing some tips. I am not an expert guru but I followed along and below are the outputs I got back: (Still not working). I did try: salt securityonion_standalone.sls state.apply idstools 

[netadmin@securityonion minions]$ **sudo systemctl status salt-master** 
â salt-master.service - The Salt Master Server
   Loaded: loaded (/usr/lib/systemd/system/salt-master.service; enabled; vendor preset: disabled)
   Active: active (running) since Wed 2021-01-13 19:17:31 UTC; 23h ago
     Docs: man:salt-master(1)
           file:///usr/share/doc/salt/html/contents.html
           https://docs.saltstack.com/en/latest/contents.html
 Main PID: 1648 (salt-master)
    Tasks: 43
   Memory: 357.6M
   CGroup: /system.slice/salt-master.service
           â1648 /usr/bin/python3 /usr/bin/salt-master
           â2076 /usr/bin/python3 /usr/bin/salt-master
           â2273 /usr/bin/python3 /usr/bin/salt-master
           â2276 /usr/bin/python3 /usr/bin/salt-master
           â2277 /usr/bin/python3 /usr/bin/salt-master
           â2278 /usr/bin/python3 /usr/bin/salt-master
           â2280 /usr/bin/python3 /usr/bin/salt-master
           â?2282 /usr/bin/python3 /usr/bin/salt-master
           ├─2289 /usr/bin/python3 /usr/bin/salt-master
           ├─2290 /usr/bin/python3 /usr/bin/salt-master
           ├─2291 /usr/bin/python3 /usr/bin/salt-master
           ├─2292 /usr/bin/python3 /usr/bin/salt-master
           ├─2297 /usr/bin/python3 /usr/bin/salt-master
           └─2306 /usr/bin/python3 /usr/bin/salt-master

Jan 14 19:10:07 securityonion salt-master[1648]: During handling of the above exception, another exception occurred:
Jan 14 19:10:07 securityonion salt-master[1648]: Traceback (most recent call last):
Jan 14 19:10:07 securityonion salt-master[1648]: File "/usr/lib/python3.6/site-packages/salt/pillar/__init__.py", lin...state
Jan 14 19:10:07 securityonion salt-master[1648]: **defaults
Jan 14 19:10:07 securityonion salt-master[1648]: File "/usr/lib/python3.6/site-packages/salt/template.py", line 101, ...plate
Jan 14 19:10:07 securityonion salt-master[1648]: ret = render(input_data, saltenv, sls, **render_kwargs)
Jan 14 19:10:07 securityonion salt-master[1648]: File "/usr/lib/python3.6/site-packages/salt/renderers/yaml.py", line...ender
Jan 14 19:10:07 securityonion salt-master[1648]: raise SaltRenderError(err_type, line_num, exc.problem_mark.buffer)
Jan 14 19:10:07 securityonion salt-master[1648]: salt.exceptions.SaltRenderError: found character that cannot start any token
Jan 14 19:10:07 securityonion salt-master[1648]: [CRITICAL] Pillar render error: Rendering SLS 'minions.securityonion...ails.
Hint: Some lines were ellipsized, use -l to show in full.

------------------------------------------------------------------------------------------------------------------------------

[netadmin@securityonion minions]$ **sudo systemctl restart salt-master**

[netadmin@securityonion minions]$ **sudo systemctl status salt-master**  
● salt-master.service - The Salt Master Server
   Loaded: loaded (/usr/lib/systemd/system/salt-master.service; enabled; vendor preset: disabled)
   Active: active (running) since Thu 2021-01-14 19:14:56 UTC; 20s ago
     Docs: man:salt-master(1)
           file:///usr/share/doc/salt/html/contents.html
           https://docs.saltstack.com/en/latest/contents.html
 Main PID: 10684 (salt-master)
    Tasks: 43
   Memory: 216.7M
   CGroup: /system.slice/salt-master.service
           ├─10684 /usr/bin/python3 /usr/bin/salt-master
           ├─10779 /usr/bin/python3 /usr/bin/salt-master
           ├─10795 /usr/bin/python3 /usr/bin/salt-master
           ├─10798 /usr/bin/python3 /usr/bin/salt-master
           ├─10799 /usr/bin/python3 /usr/bin/salt-master
           ├─10800 /usr/bin/python3 /usr/bin/salt-master
           ├─10801 /usr/bin/python3 /usr/bin/salt-master
           ├─10802 /usr/bin/python3 /usr/bin/salt-master
           ├─10809 /usr/bin/python3 /usr/bin/salt-master
           ├─10810 /usr/bin/python3 /usr/bin/salt-master
           ├─10811 /usr/bin/python3 /usr/bin/salt-master
           ├─10813 /usr/bin/python3 /usr/bin/salt-master
           ├─10814 /usr/bin/python3 /usr/bin/salt-master
           └─10820 /usr/bin/python3 /usr/bin/salt-master

Jan 14 19:15:06 securityonion salt-master[10684]: During handling of the above exception, another exception occurred:
Jan 14 19:15:06 securityonion salt-master[10684]: Traceback (most recent call last):
Jan 14 19:15:06 securityonion salt-master[10684]: File "/usr/lib/python3.6/site-packages/salt/pillar/__init__.py", line 884, in ren...pstate
Jan 14 19:15:06 securityonion salt-master[10684]: **defaults
Jan 14 19:15:06 securityonion salt-master[10684]: File "/usr/lib/python3.6/site-packages/salt/template.py", line 101, in compile_template
Jan 14 19:15:06 securityonion salt-master[10684]: ret = render(input_data, saltenv, sls, **render_kwargs)
Jan 14 19:15:06 securityonion salt-master[10684]: File "/usr/lib/python3.6/site-packages/salt/renderers/yaml.py", line 72, in render
Jan 14 19:15:06 securityonion salt-master[10684]: raise SaltRenderError(err_type, line_num, exc.problem_mark.buffer)
Jan 14 19:15:06 securityonion salt-master[10684]: salt.exceptions.SaltRenderError: found character that cannot start any token
Jan 14 19:15:06 securityonion salt-master[10684]: [CRITICAL] Pillar render error: Rendering SLS 'minions.securityonion_standalone' ...tails.
Hint: Some lines were ellipsized, use -l to show in full.

------------------------------------------------------------------------------------------------------------------------------

[netadmin@securityonion minions]$ **salt securityonion_standalone.sls state.apply idstools**
[ERROR   ] 
Salt request timed out. The master is not responding. You may need to run your command with `--async` in order to bypass the congested event bus. With `--async`, the CLI tool will print the job id (jid) and exit immediately without listening for responses. You can then use `salt-run jobs.lookup_jid` to look up the results of the job in the job cache later.

------------------------------------------------------------------------------------------------------------------------------

[netadmin@securityonion minions]$ **salt * test.ping**
[ERROR   ] 
Salt request timed out. The master is not responding. You may need to run your command with `--async` in order to bypass the congested event bus. With `--async`, the CLI tool will print the job id (jid) and exit immediately without listening for responses. You can then use `salt-run jobs.lookup_jid` to look up the results of the job in the job cache later.

------------------------------------------------------------------------------------------------------------------------------

[netadmin@securityonion minions]$ **salt * test.ping --async**
[ERROR   ] An un-handled exception was caught by salt's global exception handler:
AuthenticationError: Authentication error occurred.
Traceback (most recent call last):
  File "/usr/bin/salt", line 11, in <module>
    load_entry_point('salt==3002.2', 'console_scripts', 'salt')()
  File "/usr/lib/python3.6/site-packages/salt/scripts.py", line 539, in salt_main
    client.run()
  File "/usr/lib/python3.6/site-packages/salt/cli/salt.py", line 151, in run
    jid = self.local_client.cmd_async(**kwargs)
  File "/usr/lib/python3.6/site-packages/salt/client/__init__.py", line 447, in cmd_async
    tgt, fun, arg, tgt_type, ret, jid=jid, kwarg=kwarg, listen=False, **kwargs
  File "/usr/lib/python3.6/site-packages/salt/client/__init__.py", line 348, in run_job
    **kwargs
  File "/usr/lib/python3.6/site-packages/salt/client/__init__.py", line 1873, in pub
    raise AuthenticationError(err_msg)
salt.exceptions.AuthenticationError: Authentication error occurred.
Traceback (most recent call last):
  File "/usr/bin/salt", line 11, in <module>
    load_entry_point('salt==3002.2', 'console_scripts', 'salt')()
  File "/usr/lib/python3.6/site-packages/salt/scripts.py", line 539, in salt_main
    client.run()
  File "/usr/lib/python3.6/site-packages/salt/cli/salt.py", line 151, in run
    jid = self.local_client.cmd_async(**kwargs)
  File "/usr/lib/python3.6/site-packages/salt/client/__init__.py", line 447, in cmd_async
    tgt, fun, arg, tgt_type, ret, jid=jid, kwarg=kwarg, listen=False, **kwargs
  File "/usr/lib/python3.6/site-packages/salt/client/__init__.py", line 348, in run_job
    **kwargs
  File "/usr/lib/python3.6/site-packages/salt/client/__init__.py", line 1873, in pub
    raise AuthenticationError(err_msg)
salt.exceptions.AuthenticationError: Authentication error occurred.
[netadmin@securityonion minions]$

**Comment by TOoSmOotH** [comment]
idstools only run on the manager. What happen when you run sudo salt-call state.apply idstools on the manager?

**Reply by jlaboy71** [reply]
Hi, thanks for responding.

this is the output: 

[netadmin@securityonion /]$ sudo salt-call state.apply 
local:
    Data failed to compile:
----------
    Pillar failed to render with the following messages:
----------
    Rendering SLS 'minions.securityonion_standalone' failed. Please see master log for details.
[netadmin@securityonion /]$

**Comment by m0duspwnens** [comment, marked as answer]
There is some formatting issue with your minion file located at
/opt/so/saltstack/local/minions/securityonion_standalone.sls

**Reply by jlaboy71** [reply]
Thanks for all the responses. It seems that something isn't right. I tried to revert my only two lines that I added under the **idstools:** section under the subfield **disabled:** I figured that there might have been something about what you said as I had used a tab to type in both SIDs - 2016149 and - 2016150 so I removed those lines and added space bar spaces then saved and attempted the command: **salt securityonion_standalone.sls state.apply idstools**
It still responded with the error: Salt request timed out. The master is not responding. You may need to run your command with `--async` ...

So I then tried to just removed those two disabled SIDs and reverted back to the original file. I am still getting the same sad results.

I do have a question. Say I added those disabled SIDs, saved the file, and then rebooted the entire server, would that work or is there any other way to restart the services that process this file and applies the new changes other than the command recommended?

BTW, when I do add the --async it still outputs the error: An un-handled exception was caught by salt's global exception handler: AuthenticationError: Authentication error occurred.

Thank you very much for taking the time to read this and for offering your knowledge. I really do appreciate it. 

Background: The Security Onion install was done so using all the default settings during the installation process menu prompts and is currently running on its own independent physical box. 

Jose

**Reply by m0duspwnens** [reply]
@jlaboy71 
There are a couple issues going on here.

When you run `salt securityonion_standalone.sls state.apply idstools`, this command must be run as root or using sudo. Also, when running this command, `securityonion_standalone.sls` should just be `securityonion_standalone`. The format of the command is `sudo salt <target> state.apply <state_to_apply>`. In this case, your target is the node with the minion id `securityonion_standalone` and the state being applied is the `idstools` state. So the command would look like `sudo salt securityonion_standalone state.apply idstools`.

Now, when you run that, if you still receive the error `Rendering SLS 'minions.securityonion_standalone' failed. Please see master log for details.`, then you need to check that the minion pillar file located at `/opt/so/saltstack/local/minions/securityonion_standalone.sls` is properly formatted yaml.

-Josh

**Reply by jlaboy71** [reply]
Josh,

Many thanks for your guidance! I believe that did it. I'll just have to give it a bit to make sure I don't see those specific SIDs being reported. For technical information for the next guy who might run into this, I will include below the output of my attempt. It reported that the PID was in Queue so I had to end that task.

**Technical Details:
Note: The first attempt at this command is on the original unmodified file.**

[netadmin@securityonion /]$ sudo salt securityonion_standalone state.apply idstools
securityonion_standalone:
    Data failed to compile:
----------
    The function "state.apply" is running as PID 25791 and was started at 2021, Jan 14 21:51:35.234220 with jid 20210114215135234220
ERROR: Minions returned with non-zero exit code

**Note: I killed the process and rerun the command.**

[netadmin@securityonion /]$ sudo kill -9 25791
[netadmin@securityonion /]$ sudo salt securityonion_standalone state.apply idstools

securityonion_standalone:
----------
          ID: idstoolsdir
    Function: file.directory
        Name: /opt/so/conf/idstools/etc
      Result: True
     Comment: The directory /opt/so/conf/idstools/etc is in the correct state
     Started: 20:37:25.815110
    Duration: 29.863 ms
     Changes:   
----------
          ID: idstoolslogdir
    Function: file.directory
        Name: /opt/so/log/idstools
      Result: True
     Comment: The directory /opt/so/log/idstools is in the correct state
     Started: 20:37:25.845199
    Duration: 2.431 ms
     Changes:   
----------
          ID: idstoolsetcsync
    Function: file.recurse
        Name: /opt/so/conf/idstools/etc
      Result: True
     Comment: The directory /opt/so/conf/idstools/etc is in the correct state
     Started: 20:37:25.847798
    Duration: 155.277 ms
     Changes:   
----------
          ID: so-ruleupdatecron
    Function: cron.present
        Name: /usr/sbin/so-rule-update > /opt/so/log/idstools/download.log 2>&1
      Result: True
     Comment: Cron /usr/sbin/so-rule-update > /opt/so/log/idstools/download.log 2>&1 already present
     Started: 20:37:26.018235
    Duration: 15.931 ms
     Changes:   
----------
          ID: rulesdir
    Function: file.directory
        Name: /opt/so/rules/nids
      Result: True
     Comment: The directory /opt/so/rules/nids is in the correct state
     Started: 20:37:26.034497
    Duration: 3.038 ms
     Changes:   
----------
          ID: synclocalnidsrules
    Function: file.recurse
        Name: /opt/so/rules/nids/
      Result: True
     Comment: The directory /opt/so/rules/nids/ is in the correct state
     Started: 20:37:26.037720
    Duration: 79.749 ms
     Changes:   
----------
          ID: so-idstools
    Function: docker_container.running
      Result: True
     Comment: Container 'so-idstools' is already configured as specified
     Started: 20:37:26.434551
    Duration: 3206.084 ms
     Changes:   
----------
          ID: append_so-idstools_so-status.conf
    Function: file.append
        Name: /opt/so/conf/so-status/so-status.conf
      Result: True
     Comment: File /opt/so/conf/so-status/so-status.conf is in correct state
     Started: 20:37:29.641210
    Duration: 40.56 ms
     Changes:   

Summary for securityonion_standalone
------------
Succeeded: 8
Failed:    0
------------
Total states run:     8
Total run time:   3.533 s
==================================================================

**Note: It runs successfully so I then go to the file and add the two SIDs I want to disable and attempt to rerun the command.** 

[netadmin@securityonion /]$ sudo nano /opt/so/saltstack/local/pillar/minions/securityonion_standalone.sls
[netadmin@securityonion /]$ sudo salt securityonion_standalone state.apply idstools                      
securityonion_standalone:
    Minion did not return. [No response]
    The minions may not have all finished running and any remaining minions will return upon completion. To look up the return data for this job later, run the following command:
    
    salt-run jobs.lookup_jid 20210115203954701358
ERROR: Minions returned with non-zero exit code

**Note: I got the error above so I tried to rerun again and see the PID so I kill the PID and try rerunning the command again and it takes the second time:**

[netadmin@securityonion /]$ sudo salt securityonion_standalone state.apply idstools                      
securityonion_standalone:
    Data failed to compile:
----------
    The function "state.highstate" is running as PID 20714 and was started at  with jid req
ERROR: Minions returned with non-zero exit code
[netadmin@securityonion /]$ sudo kill -9 20714
[netadmin@securityonion /]$ sudo salt securityonion_standalone state.apply idstools
securityonion_standalone:
----------
          ID: idstoolsdir
    Function: file.directory
        Name: /opt/so/conf/idstools/etc
      Result: True
     Comment: The directory /opt/so/conf/idstools/etc is in the correct state
     Started: 20:41:30.552579
    Duration: 12.31 ms
     Changes:   
----------
          ID: idstoolslogdir
    Function: file.directory
        Name: /opt/so/log/idstools
      Result: True
     Comment: The directory /opt/so/log/idstools is in the correct state
     Started: 20:41:30.565073
    Duration: 2.134 ms
     Changes:   
----------
          ID: idstoolsetcsync
    Function: file.recurse
        Name: /opt/so/conf/idstools/etc
      Result: True
     Comment: The directory /opt/so/conf/idstools/etc is in the correct state
     Started: 20:41:30.567363
    Duration: 141.395 ms
     Changes:   
----------
          ID: so-ruleupdatecron
    Function: cron.present
        Name: /usr/sbin/so-rule-update > /opt/so/log/idstools/download.log 2>&1
      Result: True
     Comment: Cron /usr/sbin/so-rule-update > /opt/so/log/idstools/download.log 2>&1 already present
     Started: 20:41:30.713328
    Duration: 18.051 ms
     Changes:   
----------
          ID: rulesdir
    Function: file.directory
        Name: /opt/so/rules/nids
      Result: True
     Comment: The directory /opt/so/rules/nids is in the correct state
     Started: 20:41:30.731844
    Duration: 4.054 ms
     Changes:   
----------
          ID: synclocalnidsrules
    Function: file.recurse
        Name: /opt/so/rules/nids/
      Result: True
     Comment: The directory /opt/so/rules/nids/ is in the correct state
     Started: 20:41:30.736145
    Duration: 63.367 ms
     Changes:   
----------
          ID: so-idstools
    Function: docker_container.running
      Result: True
     Comment: Container 'so-idstools' is already configured as specified. State changed from 'stopped' to 'running'.
     Started: 20:41:30.906149
    Duration: 6185.626 ms
     Changes:   
              ----------
              state:
                  ----------
                  new:
                      running
                  old:
                      stopped
----------
          ID: append_so-idstools_so-status.conf
    Function: file.append
        Name: /opt/so/conf/so-status/so-status.conf
      Result: True
     Comment: File /opt/so/conf/so-status/so-status.conf is in correct state
     Started: 20:41:37.092298
    Duration: 11.185 ms
     Changes:   

Summary for securityonion_standalone
------------
Succeeded: 8 (changed=1)
Failed:    0
------------
Total states run:     8
Total run time:   6.438 s

**Comment by m0duspwnens** [comment]
@jlaboy71 
When you see the output `The function "state.apply" is running as PID` or `The function "state.highstate" is running as PID`, this doesn't  indicate any serious error. It just means that a state is currently running and the one you just tried to run can't run right now. This message could be avoided by running `sudo salt <target> state.apply <state_to_apply> queue=True` from the manager or `sudo salt-call state.apply <state_to_apply> queue=True` from the remote node.

**Comment by rwwiv** [comment]
@jlaboy71 
We're currently in the process of writing a tool to make disabling SIDs easier. If you'd like to test it out early, copy [this link's content](https://raw.githubusercontent.com/Security-Onion-Solutions/securityonion/feature/so-rules/salt/common/tools/sbin/so-rules) into `/usr/sbin/so-rules` on the manager (or alternatively, copy the script from the `feature/so-rules` branch of this repo - found under `securityonion/salt/common/tools/sbin`), then copy the following script somewhere onto the same system and run it as root:
``` sh
#!/bin/bash

. /usr/sbin/so-common

# Surround double curly braces in any minion pillar in single quotes so the pillars can be parsed as regular yaml

readarray -t minion_pillars <<< "$(find /opt/so/saltstack/local/pillar/minions -type f -name '*.sls')"
for pillar in "${minion_pillars[@]}"; do
  sed -i -r "s/ (\{\{.*}})$/ '\1'/g" "$pillar" 
done
```
The tool has built in help with `-h` to assist you with syntax.

**Reply by jlaboy71** [reply]
Thanks RWWIV,

I am currently following up on this as time has lapsed in business on other things. I tried to click the link provided but it doesn't go anywhere.

Jose
